ably/proto_protocol_message_test.go

```diff
@@ -29,6 +30,82 @@ func TestProtocolMessageEncodeZeroSerials(t *testing.T) {
 		"unexpected msgpack encoding\nexpected: %x\nactual:   %x", expected, encoded)
 }
 
+func TestUpdateEmptyMessageFields_TM2a_TM2c_TM2f(t *testing.T) {
+	messages := []*ably.Message{
+		{
+			ID:           "",
+			ConnectionID: "",
+			Timestamp:    0,
+		},
+		{
+			ID:           "",
+			ConnectionID: "",
+			Timestamp:    0,
+		},
+		{
+			ID:           "",
+			ConnectionID: "",
+			Timestamp:    0,
+		},
+	}
+
+	presenceMessages := []*ably.PresenceMessage{
+		{
+			Message: ably.Message{
+				ID:           "",
+				ConnectionID: "",
+				Timestamp:    0,
+			},
+			Action: 0,
+		},
+		{
+			Message: ably.Message{
+				ID:           "",
+				ConnectionID: "",
+				Timestamp:    0,
+			},
+			Action: 0,
+		},
+		{
+			Message: ably.Message{
+				ID:           "",
+				ConnectionID: "",
+				Timestamp:    0,
+			},
+			Action: 0,
+		},
+		{
+			Message: ably.Message{
+				ID:           "",
+				ConnectionID: "",
+				Timestamp:    0,
+			},
+			Action: 0,
+		},
+	}
+	protoMsg := ably.ProtocolMessage{
+		Messages:     messages,
+		Presence:     presenceMessages,
+		ID:           "msg-id",
+		ConnectionID: "conn-id",
+		Timestamp:    3453,
+	}
+
+	protoMsg.UpdateEmptyFields()
+
+	for msgIndex, msg := range protoMsg.Messages {
+		assert.Equal(t, protoMsg.ID+":"+strconv.Itoa(msgIndex), msg.ID)
+		assert.Equal(t, protoMsg.ConnectionID, msg.ConnectionID)
+		assert.Equal(t, protoMsg.Timestamp, msg.Timestamp)
+	}
+
+	for presenceMsgIndex, presenceMessage := range protoMsg.Presence {
+		assert.Equal(t, protoMsg.ID+":"+strconv.Itoa(presenceMsgIndex), presenceMessage.ID)
+		assert.Equal(t, protoMsg.ConnectionID, presenceMessage.ConnectionID)
+		assert.Equal(t, protoMsg.Timestamp, presenceMessage.Timestamp)
+	}
```

Comment:
```suggestion
	for msgIndex, msg := range protoMsg.Messages {
		assert.Equal(t, "msg-id:"+strconv.Itoa(msgIndex), msg.ID)
		assert.Equal(t, "conn-id", msg.ConnectionID)
		assert.Equal(t, int64(3453), msg.Timestamp)
	}

	for presenceMsgIndex, presenceMessage := range protoMsg.Presence {
		assert.Equal(t, "msg-id:"+strconv.Itoa(presenceMsgIndex), presenceMessage.ID)
		assert.Equal(t, "conn-id", presenceMessage.ConnectionID)
		assert.Equal(t, int64(3453), presenceMessage.Timestamp)
	}
```
For test assertions, it's good practice to hard code the first value (the expected value) rather than calculating it or using variable to represent it. This increases test readability and guards against accidentally writing tautological tests ( https://randycoulman.com/blog/2016/12/20/tautological-tests/ )